### PR TITLE
Support yanking to system clipboard 

### PR DESCRIPTION
**Pull request by CBenoit**
Implements register selection with `"` command and system clipboard yanking.

This is WIP because I can't get `copypasta` to work on my machine. Their example is not even working on my machine.
I guess the issue would be on `copypasta` side.

Let me know what you think.

### CODE REVIEW

**Comment by archseer** [comment]
I think this looks good! What platform are you testing on? I think the issue is likely that `copypasta` defaults to `NOP` unless a specific feature is chosen: https://github.com/alacritty/copypasta/blob/f4125b09c30bee525327660e27411b76ddbe0718/src/lib.rs#L50-L77

Alacritty itself has some extra code to pick the provider, we might want to copy that directly: https://github.com/alacritty/alacritty/blob/531e494cf9a90de91dbbb84cb6cfc3158b78f1f2/alacritty/src/clipboard.rs#L8-L63

Also, feel free to join us at https://matrix.to/#/#helix-community:matrix.org to discuss (make sure to join #helix-editor:matrix.org if you're on a client that doesn't support Matrix Spaces yet)

**Comment by CBenoit** [comment]
I'm running on Linux (at the risk of participating to the "btw I use arch" meme, my distro is archlinux). I believe the right provider (X11ClipboardContext) is selected because I actually get my system clipboard with `get_contents`:
```
[helix-core/src/register.rs:43] &ClipboardContext::new()?.get_contents() = Ok("I actually get my system clipboard")
```
However, `alacritty` is my main terminal and works perfectly. I'll try to use their code and see if it's better!

File: helix-term/src/ui/editor.rs [hunk 1]
```diff
@@ -523,6 +523,9 @@ impl EditorView {
                 // if this fails, count was Some(0)
                 // debug_assert!(cxt.count != 0);
 
+                // set the register
+                cxt.register = cxt.editor.register.take().unwrap_or('"');
```
Comment (pickfire) [inline helix-term/src/ui/editor.rs:527]: We shouldn't default to `"`. I don't think by default we need to yank to system register.

Comment (archseer) [reply helix-term/src/ui/editor.rs:527]: `"` seems to be just the general register in kakoune (https://github.com/mawww/kakoune/blob/master/doc/pages/registers.asciidoc#default-registers), we're using `+` for the clipboard

Comment (pickfire) [reply helix-term/src/ui/editor.rs:527]: If it's the default why don't we just make it `char` rather than `Option<char>`?

Comment (CBenoit) [reply helix-term/src/ui/editor.rs:527]: Similar to how `count` is handled (even though default is `1`), it indicates that a value has been explicitly set with `"` command. Behavior is pretty much the same though.

**Comment by CBenoit** [comment]
Update: `alacritty` is doing what it does to support primary / secondary clipboards on x11 basically. We don't need that if all we want is the "default" system clipboard (in `+` register). However, we might use the same approach if we want to support a secondary clipboard in `*` for instance (neovim does that).

File: helix-term/src/ui/editor.rs (hunk 1 above)
Comment (pickfire) [reply helix-term/src/ui/editor.rs:527]: We could do `mem::replace` which is a single line either.

Comment (archseer) [reply helix-term/src/ui/editor.rs:527]: I do prefer the explicit `Option` on both `Editor` and the `Context` so we know if a register was explicitly set or not: that's one of the issues with `count` currently: it uses an Option on the `Editor`, but it's an `int` on the `Context` so there's problems with `1g`: https://github.com/helix-editor/helix/blob/8c2fa12ffc006de787992b4022263be033ac8200/helix-term/src/commands.rs#L1286-L1287

Comment (CBenoit) [reply helix-term/src/ui/editor.rs:527]: Maybe I should use `Option<char>` on `Context` as well then?

Comment (archseer) [reply helix-term/src/ui/editor.rs:527]: Yep!

Comment (pickfire) [reply helix-term/src/ui/editor.rs:527]: > I do prefer the explicit Option on both Editor and the Context so we know if a register was explicitly set or not: that's one of the issues with count currently: it uses an Option on the Editor, but it's an int on the Context so there's problems with 1g: 

Can we do `Option<NonZeroUsize>` for that? We can also be sure `0` won't be valid.

**Comment by pickfire** [comment]
> Update: alacritty is doing what it does to support primary / secondary clipboards on x11 basically. We don't need that if all we want is the "default" system clipboard (in + register). However, we might use the same approach if we want to support a secondary clipboard in * for instance (neovim does that).

I personally prefer yank not messing with primary clipboard (ctrl-c) unless I specified `"*`, changing selection (middle-click) is okay for me though.

**Comment by CBenoit** [comment]
> > Update: alacritty is doing what it does to support primary / secondary clipboards on x11 basically. We don't need that if all we want is the "default" system clipboard (in + register). However, we might use the same approach if we want to support a secondary clipboard in * for instance (neovim does that).
> 
> I personally prefer yank not messing with primary clipboard (ctrl-c) unless I specified `"*`, changing selection (middle-click) is okay for me though.

I think that would be the plan

**Comment by archseer** [comment]
Looks like we need to `apt install libxcb-xfixes0-dev` if `matrix.os` is ubuntu.

https://github.com/alacritty/alacritty/blob/3e867a056018c507d79396cb5c5b4b8309c609c2/.github/workflows/release.yml#L66

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -51,6 +51,9 @@ jobs:
     name: Test Suite
     runs-on: ${{ matrix.os }}
     steps:
+      - name: Install dependencies
```
Comment (archseer) [inline .github/workflows/build.yml:54]: `if: matrix.os == "ubuntu-latest"`

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -51,6 +51,9 @@ jobs:
...
+        run: sudo apt-get install libxcb-xfixes0-dev
```
Comment (archseer) [inline .github/workflows/build.yml:55]: We should add a comment to the README that `libxcb` is required to build on Linux. It should be already present on X11 systems but might be missing for Wayland users (maybe we should make `x11` a feature that's on by default so people can opt out).

File: .github/workflows/build.yml (hunk 2 above)
Comment (CBenoit) [reply .github/workflows/build.yml]: Exactly what I was looking for

Comment (archseer) [reply .github/workflows/build.yml]: Note: need to add this dependency to `release.yml` too

File: .github/workflows/build.yml (hunk 3 above)
Comment (CBenoit) [reply .github/workflows/build.yml:57]: The option makes sense for that reason. Alacritty does that as well.

**Comment by archseer** [comment]
Bad news: after testing on Wayland, the clipboard doesn't work for me. It seems that it always enters X11 mode and would need some extra work. I found a fork that would try starting in Wayland mode first, then fallback to X11.

That said, I don't like the extra runtime deps (almost 30 crates on Linux) and dynamic libraries we need to link against (it's really nice that there's no external dependencies at the moment). I think this is something [neovim got right](https://github.com/neovim/neovim/blob/f2906a4669a2eef6d7bf86a29648793d63c98949/runtime/autoload/provider/clipboard.vim#L68-L152) they removed the internal code and it simply integrates with external clipboard providers. I think we could do this easily via `exists?` and `std::process::Command`.

What do you think about extracting the register selection into a new PR, then redoing the clipboard this way? Sorry for all the hassle...

**Comment by CBenoit** [comment]
Totally agreed.

**Comment by alper** [comment]
So how do you yank to the keyboard? (Sorry but this is the only documentation that comes up.)

**Comment by CBenoit** [comment]
> So how do you yank to the keyboard? (Sorry but this is the only documentation that comes up.)

Try searching for ["system clipboard" in the documentation](https://docs.helix-editor.com/title-page.html?search=system%20clipboard)

**Pull request closed without merging**